astropy/utils/data_info.py

```diff
@@ -231,46 +306,40 @@ class DataInfo:
 
     def __init__(self, bound=False):
         # If bound to a data object instance then create the dict of attributes
-        # which stores the info attribute values.
+        # which stores the info attribute values. Default of None for "unset"
+        # except for dtype where the default is object.
         if bound:
-            self._attrs = dict((attr, None) for attr in self.attr_names)
+            self._attrs = {}
 
     @property
     def _parent(self):
-        if self._parent_ref is None:
+        try:
```

Comment:
Trying to speed this critical property up...

Reply:
Another thing to put on the table is making this a direct ref instead of weak ref to the instance.  Making the weak ref for every access to `info` takes ~100 ns.  Since this ref is contained in an object stored on the instance, it seems like this is the straightforward reference cycle that GC should have no problem detecting.  But of course that may be simplistic and memory leaks are a bit annoying.

Reply:
Yes, I'd like that, but perhaps in the spirit of keeping things as similar as possible here, shall we leave it for later? I have this memory of confusing ourselves no end about memory leaks on `info` before... (Or was that for links to parent table??)